lab7b.md

```diff
@@ -0,0 +1,141 @@
+# Lab 7b - Using secrets
+
+<img src="https://github.com/openfaas/media/raw/master/OpenFaaS_Magnet_3_1_png.png" width="500px"></img>
+
+[Lab 7](https://github.com/openfaas/workshop/blob/master/lab7.md) looked at how the GitHub *Personal Access Token* could be stored in a function enviroment variable (`auth_token`) by including it in a file called `env.yml`.  An alternative approach is to use a **secret** to store sensitive information.
+
+Docker describe a secret as: 
+> .. a blob of data, such as a password, SSH private key, SSL certificate, or another piece of data that should not be transmitted over a network or stored unencrypted in a Dockerfile or in your application’s source code.
+
+Seems a good fit for storing the `auth_token` value.
+
+## Create a secret
+
+> To be cross platform compatible use of '_' in the name of secrets should be avoided.
+
+From a terminal run the following command:
+
+```
+$ echo -n <auth_token> | docker secret create auth-token -
+```
+
+Test that the secret was created:
+
+```
+$ docker secret inspect auth-token
+```
+
+When the secret is mounted by a function it will be presented as a file under `/run/secrets/auth-token`. This can be read by `handler.py` to obtain the GitHub *Personal Access Token*.
+
+## Update issue-bot.yml
+
+Replace the reference to `env.yml` with an instruction to make the `auth-token` secret available to the function:
+
+```yml
+provider:
+  name: faas
+  gateway: http://127.0.0.1:8080
+
+functions:
+  issue-bot:
+    lang: python3
+    handler: ./issue-bot
+    image: <your-username>/issue-bot
+    environment:
+      write_debug: true
+      gateway_hostname: "gateway"
+      positive_threshold: 0.25
+    secrets:
+    - auth-token
+
+```
+
+## Update the `issue-bot` function
+
+The function handler requires changing in order to cause it to read the `auth-token` secret, rather than the environment variable.  This is a single line change where:
+
+```python
+g = Github(os.getenv("auth_token"))
+``` 
+is replaced with 
+```python
+authToken = open("/run/secrets/auth-token","r")
```

Comment:
Try to use the pattern of `with` because it ensures the file gets closed afterwards.

http://www.pythonforbeginners.com/files/reading-and-writing-files-in-python

```
with open(“testfile.txt”) as file:  
    data = file.read() 
    # do something with data 
```